### PR TITLE
 [ Task ] Implementing caching in the /users/{user_id}/tasks/{task_id} (GET, PUT, DELETE) Invalidate users:{user_id}/tasks/{task_id}’ on PUT, DELETE. 

### LINKED ISSUE #153: [ Task ]  Implementing caching in the /users/{user_id}/tasks/{task_id}
What part of the code implements the functionality? What to pull?  How to deploy the functionality locally or in the cloud? How to configure the functionality?
     - The part of the code that implements the functionality: importing redis and the r = redis.StrictRedis that connects the host URL and the port. Also the code that connects to the DB and checks if the data in the DB is in the redis server and if not then it dumps it in the Redis Server. Deploy the function locally by cd- ing into the wanted file and run the program in the top of the screen on VS click on the run without Debugging. You can configure the functionality by changing the code that pulls the data from the DB. Right now it is TaskID but it can be changed to other ID's but it would have to connect to a different function. 

What input to provide? What output is expected? Known troubleshooting steps
     - The input provided is the TaskId and the output expected is the data related to the TaskId. The problems right now are connection to the DB. Natalia's resource was deleted and the connection to the DB would not work but the connection to the Redis server does work. Last minute Natalia fixed her resource and was able to pull data testing on Azure but not fully in the coding. 
      
DATE | ACTIVITY | TIME
-- | -- | --
2/9 | Research Redis Cash and Docker | 30 mins
2/11 | Sign up and install docker but turns out I did not need docker| 1.5 hours
2/13| Group meeting and talk about Redis Cash | 2 hours
2/15 | Talk to Luca and get URL link for Redis Server Resource | 30 mins
2/15 | Start Testing and Running the connection to the URL, Connection Issues to DB | 3 hour
2/16 |Connection Issues to DB continuing, Cannot Pull data out of DB | 3 hour
  | TOTAL HOURS SPENT | 10 hours

### LINKED ISSUE #206: [Task] Implementing caching in the /users/{user_id}/tasks/{task_id}  Invalidate the cache ‘users:{user_id}:tasks:all’ for PUT and DELETE and save the GET result, invalidate the ‘users:{user_id}/tasks/{task_id}’ on PUT and DELETE where the same users IDs are used

### PR DESCRIPTION
**Pull request by jaguiniga**
Added connection to Redis Server and pull Task ID data from DB and dump in Redis Cash

### CODE REVIEW

**Review by kieranmiguel — CHANGES_REQUESTED** [review]
Remove hard-coded password

**Review by chonger878 — COMMENTED** [review]
Just curious, are you planning on making the host and port a secret/environment variable?  I know the password has to be one.

**Review by chonger878 — APPROVED** [review]

**Review by kieranmiguel — APPROVED** [review]

File: functions/HttpTriggerAPIUsersIdTaskId/__init__.py [hunk 1]
```diff
@@ -6,40 +6,66 @@
...
 import pyodbc
 import azure.functions as func
+import redis 
+
+# Connect to the Redis Server
+r = redis.StrictRedis(host='nsc-redis-dev-usw2-thursday.redis.cache.windows.net', 
+port=6380, db=0, password= '${{ secrets.ENV_REDIS_KEY }}', ssl=True)
+
+# Set Message in the Redis Server for testing
+r.set("Message01", "Hello World")
```
Comment (unicornsleuth) [inline functions/HttpTriggerAPIUsersIdTaskId/__init__.py:24]: You can delete this

**Review by unicornsleuth — CHANGES_REQUESTED** [review]
I couldn't get this to work with {{ secrets.ENV_REDIS_KEY }} but if it works for you, okay.
I did get this failure though! I think you will need to convert the datetime fields to strings to make caching work, but you may be able to find another workaround.
image

File: functions/HttpTriggerAPIUsersIdTaskId/__init__.py [hunk 2]
```diff
@@ -6,40 +6,66 @@
 import os
 import pyodbc
 import azure.functions as func
+import redis 
+
+# Connect to the Redis Server
+r = redis.StrictRedis(host='nsc-redis-dev-usw2-thursday.redis.cache.windows.net', 
+port=6380, db=0, password= '${{ secrets.ENV_REDIS_KEY }}', ssl=True)
```
Comment (toddysm) [inline functions/HttpTriggerAPIUsersIdTaskId/__init__.py]: Where this secret comes from?

Comment (jaguiniga) [reply functions/HttpTriggerAPIUsersIdTaskId/__init__.py]: I fixed it, not a secret but os.ENVIRON

**Pull request merged**